File: Copy_List_with_Random_Pointer.cpp
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * @param head: The head of linked list with a random pointer.
     * @return: A new head of a deep copy of the list.
     */
    RandomListNode *copyRandomList(RandomListNode *head) {
        // write your code here
        // RandomListNode *dummy = new RandomListNode(INT_MIN);
        // dummy->next = head;
        // while (head){
        //     auto copy_head = new RandomListNode(head->label);
        //     copy_head->next = head->next;
        //     head->next = copy_head;
        //     head = copy_head->next;
        // }
        // //copy the random poiter
        // head = dummy->next;
        // while(head){
        //     auto copy_head = head->next;
        //     if (head->random){
        //         copy_head->random = head->random->next;
        //     }
        //     head = copy_head->next;
        // }
        
        // //separate
        // head = dummy;
        // RandomListNode *copy_dummy = new RandomListNode(INT_MIN);
        // auto copy_head = copy_dummy;
        // auto p = dummy->next;
        // while (p){
        //     head->next = p;
        //     copy_head->next = p->next;
        //     p = p->next->next;
        //     head = head->next;
        //     copy_head = copy_head->next;
        // }
        // head->next = nullptr;
        // head = dummy->next;
        // return copy_dummy->next;
        //good codes using for-loop
        
        for (RandomListNode *cur = head; cur; cur = cur->next->next){
            auto copy = new RandomListNode(cur->label);
            copy->next = cur->next;
            cur->next = copy;
        }
        //copy random
        for (RandomListNode *cur = head; cur; cur = cur->next->next){
            if (cur->random){
                cur->next->random = cur->random->next;
            }
        }
        //seprate
        RandomListNode dummy(INT_MIN);
        for (RandomListNode *cur = head, *copy = &dummy; cur; cur = cur->next, copy = copy->next){
            copy->next = cur->next;
            cur->next = copy->next->next;
        }
        return dummy.next;
    }
};

File: Copy List with Random Pointer.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode:
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None
class Solution:
    # @param head: A RandomListNode
    # @return: A RandomListNode
    def copyRandomList(self, head):
        # write your code here
        
        # ## copy
        # cur = head
        # while cur:
        #     next = cur.next
        #     cur.next = RandomListNode(cur.label)
        #     cur.next.next = next
        #     cur = next
        
        # ## copy random
        # cur = head
        # while cur:
        #     if cur.random:
        #         cur.next.random = cur.random.next
        #     cur = cur.next.next
            
        # ## sperate
        # cur, ans = head, head.next
        # ansCur = ans
        # while cur:
        #     cur.next, cur = cur.next.next, cur.next.next
        #     if ansCur.next:
        #         ansCur.next, ansCur = ansCur.next.next, ansCur.next.next
        # return ans
        
        dummy = RandomListNode(0)
        newCur, cur = dummy, head
        dict = {}
        while cur:
            newCur.next = RandomListNode(cur.label)
            newCur = newCur.next
            dict[cur] = newCur
            cur = cur.next
        
        ## copy random
        cur = head
        newCur = dummy.next
        while cur:
            if cur.random:
                newCur.random = dict[cur.random]
            cur, newCur = cur.next, newCur.next
        return dummy.next
        
        